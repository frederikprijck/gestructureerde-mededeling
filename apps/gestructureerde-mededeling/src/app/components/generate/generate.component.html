<ogm-number [ogm]="ogm.numberFormat"></ogm-number>

<ng-container *ngIf="isSmallScreen$ | async; then smallControls; else largeControls"></ng-container>

<ng-template #smallControls>
  <div class="controls">
    <button mat-fab disableRipple color="primary" type="button" title="Verniewen" (click)="refresh()">
      <fa-icon [icon]="['fad', 'sync-alt']"></fa-icon>
    </button>
    <button
      class="ogm-copy"
      mat-fab
      disableRipple
      color="primary"
      type="button"
      title="Kopieer nummer"
      data-clipboard-action="copy"
      [attr.data-clipboard-text]="ogm.number"
      (click)="copyToClipboard('Nummer gekopieerd naar klembord')"
    >
      <fa-icon [icon]="['fad', 'comment-dots']"></fa-icon>
    </button>
    <button
      class="ogm-copy"
      mat-fab
      disableRipple
      color="primary"
      type="button"
      title="Kopieer mededeling"
      data-clipboard-action="copy"
      [attr.data-clipboard-text]="ogm.numberFormat"
      (click)="copyToClipboard('Mededeling gekopieerd naar klembord')"
    >
      <fa-icon [icon]="['fad', 'comment-lines']"></fa-icon>
    </button>
  </div>
</ng-template>

<ng-template #largeControls>
  <div class="controls large">
    <button mat-raised-button isRoundButton disableRipple color="primary" (click)="refresh()">
      <fa-icon [icon]="['fad', 'sync-alt']"></fa-icon>Vernieuwen
    </button>
    <button
      class="ogm-copy"
      mat-raised-button
      isRoundButton
      disableRipple
      color="primary"
      data-clipboard-action="copy"
      [attr.data-clipboard-text]="ogm.number"
      (click)="copyToClipboard('Nummer gekopieerd naar klembord')"
    >
      <fa-icon [icon]="['fad', 'comment-dots']"></fa-icon>Kopieer nummer
    </button>
    <button
      class="ogm-copy"
      mat-raised-button
      isRoundButton
      disableRipple
      color="primary"
      data-clipboard-action="copy"
      [attr.data-clipboard-text]="ogm.numberFormat"
      (click)="copyToClipboard('Mededeling gekopieerd naar klembord')"
    >
      <fa-icon [icon]="['fad', 'comment-lines']"></fa-icon>Kopieer mededeling
    </button>
  </div>
</ng-template>
