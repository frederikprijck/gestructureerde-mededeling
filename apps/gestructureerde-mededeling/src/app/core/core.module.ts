import { NgModule } from '@angular/core';
import { RoutingModule } from './routing.module';
import { MaterialModule } from './material.module';
import { ReactiveFormsModule } from '@angular/forms';

@NgModule({
  declarations: [],
  imports: [RoutingModule, MaterialModule, ReactiveFormsModule],
  exports: [RoutingModule, MaterialModule, ReactiveFormsModule],
})
export class CoreModule {}
