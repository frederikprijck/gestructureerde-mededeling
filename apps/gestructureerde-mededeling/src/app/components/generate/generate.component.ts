import { ChangeDetectionStrategy, Component, OnDestroy, OnInit } from '@angular/core';
import * as ClipboardJS from 'clipboard';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { Breakpoints } from '@angular/cdk/layout';
import { MatSnackBar } from '@angular/material/snack-bar';
import { ScreenService } from '../../core/services/screen.service';
import { OgmService } from '../../core/services/ogm.service';
import { FaIconLibrary } from '@fortawesome/angular-fontawesome';
import { faSyncAlt, faCommentDots, faCommentLines } from '@fortawesome/pro-duotone-svg-icons';
import { OgmData } from '../../core/ogm.model';

@Component({
  selector: 'ogm-generate',
  templateUrl: './generate.component.html',
  styleUrls: ['./generate.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class GenerateComponent implements OnInit, OnDestroy {
  readonly isSmallScreen$: Observable<boolean>;

  ogm: OgmData;
  clipboard: ClipboardJS;

  constructor(
    private ogmService: OgmService,
    private screenService: ScreenService,
    private snackBar: MatSnackBar,
    private faIconLibrary: FaIconLibrary
  ) {
    this.isSmallScreen$ = this.screenService
      .observerBreakpoints()
      .pipe(map((breakpoints: { [key: string]: boolean }) => breakpoints[Breakpoints.XSmall]));

    this.clipboard = new ClipboardJS('.ogm-copy');

    this.faIconLibrary.addIcons(faSyncAlt, faCommentDots, faCommentLines);
  }

  ngOnInit() {
    this.refresh();
  }

  ngOnDestroy() {
    this.clipboard.destroy();
  }

  refresh(message?: string) {
    this.showSnackbarMessage(message);

    const ogmNumber = this.ogmService.generate();
    const ogmNumberFormat = this.ogmService.format(ogmNumber);

    this.ogm = { number: ogmNumber, numberFormat: ogmNumberFormat };
  }

  copyToClipboard(message: string) {
    this.showSnackbarMessage(message);
  }

  showSnackbarMessage(message?: string) {
    if (message) {
      this.snackBar.open(message, '', {
        panelClass: 'center-clipboard-text',
        verticalPosition: 'top',
        duration: 750,
      });
    }
  }
}
