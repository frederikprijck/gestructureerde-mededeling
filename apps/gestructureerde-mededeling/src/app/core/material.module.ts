import { NgModule } from '@angular/core';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { LayoutModule } from '@angular/cdk/layout';
import { MatIconModule } from '@angular/material/icon';
import { MatCardModule } from '@angular/material/card';
import { MatButtonModule } from '@angular/material/button';
import { MatSnackBarModule } from '@angular/material/snack-bar';
import { MatToolbarModule } from '@angular/material/toolbar';

@NgModule({
  declarations: [],
  imports: [
    BrowserAnimationsModule,
    MatButtonModule,
    MatCardModule,
    MatIconModule,
    LayoutModule,
    MatSnackBarModule,
    MatToolbarModule,
  ],
  exports: [MatButtonModule, MatCardModule, MatIconModule, LayoutModule, MatSnackBarModule, MatToolbarModule],
})
export class MaterialModule {}
