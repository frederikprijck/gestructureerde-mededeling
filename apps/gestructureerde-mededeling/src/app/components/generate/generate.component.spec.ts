import { async, ComponentFixture, TestBed } from '@angular/core/testing';

import { GenerateComponent } from './generate.component';
import { OgmService } from '../../core/services/ogm.service';
import { MaterialModule } from '../../core/material.module';
import { MatSnackBar, MatSnackBarConfig } from '@angular/material/snack-bar';
import { ScreenService } from '../../core/services/screen.service';
import { Component, Input } from '@angular/core';
import { FontAwesomeModule } from '@fortawesome/angular-fontawesome';

@Component({
  selector: 'ogm-number',
  template: `
    <div>{{ ogm | json }}</div>
    <div>{{ isValid | json }}</div>
  `,
})
class MockNumberComponent {
  @Input() ogm: string;
  @Input() isValid: boolean;
}

@Component({
  selector: 'ogm-input',
  template: `
    <div>{{ validate | json }}</div>
    <div>{{ placeholderMessage | json }}</div>
  `,
})
class MockInputComponent {
  @Input() validate: boolean;
  @Input() placeholderMessage: string;
}

describe('GenerateComponent', () => {
  let component: GenerateComponent;
  let fixture: ComponentFixture<GenerateComponent>;
  let snackBar: MatSnackBar;
  let ogmService: OgmService;

  const number: string = '123456789012';
  const numberFormat: string = '+++123/4567/89012+++';

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [MaterialModule, FontAwesomeModule],
      declarations: [GenerateComponent, MockNumberComponent, MockInputComponent],
      providers: [OgmService, ScreenService, MatSnackBar],
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(GenerateComponent);
    component = fixture.componentInstance;

    snackBar = TestBed.inject(MatSnackBar);
    ogmService = TestBed.inject(OgmService);

    jest.spyOn(ogmService, 'generate').mockReturnValueOnce(number);
    jest.spyOn(ogmService, 'format').mockReturnValueOnce(numberFormat);
  });

  afterEach(() => {
    jest.resetAllMocks();
    jest.restoreAllMocks();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  describe('rendering component', () => {
    it('should match snapshot', () => {
      // act
      fixture.detectChanges();

      // check
      expect(fixture).toMatchSnapshot();
    });
  });

  describe('ngOnInit', () => {
    it('should call refresh method', () => {
      // prepare
      jest.spyOn(component, 'refresh');

      // act
      component.ngOnInit();

      // check
      expect(component.refresh).toHaveBeenCalled();
    });
  });

  describe('ngOnDestroy', () => {
    it('should call destroy method', () => {
      // prepare
      jest.spyOn(component.clipboard, 'destroy');

      // act
      component.ngOnDestroy();

      // check
      expect(component.clipboard.destroy).toHaveBeenCalled();
    });
  });

  describe('refresh', () => {
    beforeEach(() => {
      jest.spyOn(component, 'showSnackbarMessage');
    });

    it('should refresh the current OGM number and call showSnackbarMessage', () => {
      // prepare
      const message: string = 'Refresh!';
      const expectedResult = {
        number,
        numberFormat,
      };

      // act
      component.refresh(message);

      // check
      expect(component.showSnackbarMessage).toHaveBeenCalledWith(message);
      expect(ogmService.generate).toHaveBeenCalled();
      expect(ogmService.format).toHaveBeenCalled();
      expect(component.ogm).toStrictEqual(expectedResult);
    });
  });

  describe('copyToClipboard', () => {
    it('should call showSnackbarMessage method', () => {
      // prepare
      jest.spyOn(component, 'showSnackbarMessage');
      const message: string = 'Copy To Clipboard!';

      // act
      component.copyToClipboard(message);

      // check
      expect(component.showSnackbarMessage).toHaveBeenCalledWith(message);
    });
  });

  describe('showSnackbarMessage', () => {
    const snackBarConfig: MatSnackBarConfig = {
      panelClass: 'center-clipboard-text',
      verticalPosition: 'top',
      duration: 750,
    };

    beforeEach(() => {
      jest.spyOn(snackBar, 'open');
    });

    it('should call open when a message is provided', () => {
      // prepare
      const message: string = 'Hello World!';

      // act
      component.showSnackbarMessage(message);

      // check
      expect(snackBar.open).toHaveBeenCalledWith(message, '', snackBarConfig);
    });

    it('should NOT call open when no message is provided', () => {
      // prepare
      const message: string = '';

      // act
      component.showSnackbarMessage(message);

      // check
      expect(snackBar.open).not.toHaveBeenCalled();
    });
  });
});
