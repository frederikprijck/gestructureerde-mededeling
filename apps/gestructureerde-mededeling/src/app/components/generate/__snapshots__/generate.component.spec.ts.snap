// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenerateComponent rendering component should match snapshot 1`] = `
<ogm-generate
  clipboard={[Function Clipboard]}
  faIconLibrary={[Function FaIconLibrary]}
  isSmallScreen$={[Function Observable]}
  ogm={[Function Object]}
  ogmService={[Function OgmService]}
  screenService={[Function ScreenService]}
  snackBar={[Function MatSnackBar]}
>
  <ogm-number>
    <div>
      "+++123/4567/89012+++"
    </div>
    <div>
      
    </div>
  </ogm-number><div
    class="controls large"
  >
    <button
      class="mat-focus-indicator mat-raised-button mat-button-base mat-primary"
      color="primary"
      disableripple=""
      isroundbutton=""
      mat-raised-button=""
    >
      <span
        class="mat-button-wrapper"
      >
        <fa-icon
          class="ng-fa-icon"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-sync-alt fa-w-16"
            data-icon="sync-alt"
            data-prefix="fad"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              class="fa-group"
            >
              <path
                class="fa-secondary"
                d="M8 454.06V320a24 24 0 0 1 24-24h134.06c21.38 0 32.09 25.85 17 41l-41.75 41.75A166.82 166.82 0 0 0 256.16 424c77.41-.07 144.31-53.14 162.78-126.85a12 12 0 0 1 11.65-9.15h57.31a12 12 0 0 1 11.81 14.18C478.07 417.08 377.19 504 256 504a247.14 247.14 0 0 1-171.31-68.69L49 471c-15.15 15.15-41 4.44-41-16.94z"
                fill="currentColor"
              />
              <path
                class="fa-primary"
                d="M12.3 209.82C33.93 94.92 134.81 8 256 8a247.14 247.14 0 0 1 171.31 68.69L463 41c15.12-15.12 41-4.41 41 17v134a24 24 0 0 1-24 24H345.94c-21.38 0-32.09-25.85-17-41l41.75-41.75A166.8 166.8 0 0 0 255.85 88c-77.46.07-144.33 53.18-162.79 126.85A12 12 0 0 1 81.41 224H24.1a12 12 0 0 1-11.8-14.18z"
                fill="currentColor"
              />
            </g>
          </svg>
        </fa-icon>
        Vernieuwen 
      </span>
      <div
        class="mat-button-ripple mat-ripple"
        matripple=""
      />
      <div
        class="mat-button-focus-overlay"
      />
    </button>
    <button
      class="ogm-copy mat-focus-indicator mat-raised-button mat-button-base mat-primary"
      color="primary"
      data-clipboard-action="copy"
      data-clipboard-text="123456789012"
      disableripple=""
      isroundbutton=""
      mat-raised-button=""
    >
      <span
        class="mat-button-wrapper"
      >
        <fa-icon
          class="ng-fa-icon"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-comment-dots fa-w-16"
            data-icon="comment-dots"
            data-prefix="fad"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              class="fa-group"
            >
              <path
                class="fa-secondary"
                d="M256 32C114.6 32 0 125.1 0 240c0 49.6 21.4 95 57 130.7C44.5 421.1 2.7 466 2.2 466.5a8 8 0 0 0-1.5 8.7A7.83 7.83 0 0 0 8 480c66.3 0 116-31.8 140.6-51.4A305 305 0 0 0 256 448c141.4 0 256-93.1 256-208S397.4 32 256 32zM128 272a32 32 0 1 1 32-32 32 32 0 0 1-32 32zm128 0a32 32 0 1 1 32-32 32 32 0 0 1-32 32zm128 0a32 32 0 1 1 32-32 32 32 0 0 1-32 32z"
                fill="currentColor"
              />
              <path
                class="fa-primary"
                d="M128 208a32 32 0 1 0 32 32 32 32 0 0 0-32-32zm128 0a32 32 0 1 0 32 32 32 32 0 0 0-32-32zm128 0a32 32 0 1 0 32 32 32 32 0 0 0-32-32z"
                fill="currentColor"
              />
            </g>
          </svg>
        </fa-icon>
        Kopieer nummer 
      </span>
      <div
        class="mat-button-ripple mat-ripple"
        matripple=""
      />
      <div
        class="mat-button-focus-overlay"
      />
    </button>
    <button
      class="ogm-copy mat-focus-indicator mat-raised-button mat-button-base mat-primary"
      color="primary"
      data-clipboard-action="copy"
      data-clipboard-text="+++123/4567/89012+++"
      disableripple=""
      isroundbutton=""
      mat-raised-button=""
    >
      <span
        class="mat-button-wrapper"
      >
        <fa-icon
          class="ng-fa-icon"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-comment-lines fa-w-16"
            data-icon="comment-lines"
            data-prefix="fad"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              class="fa-group"
            >
              <path
                class="fa-secondary"
                d="M256 32C114.6 32 0 125.1 0 240c0 49.6 21.4 95 57 130.7C44.5 421.1 2.7 466 2.2 466.5a8 8 0 0 0-1.5 8.7A7.83 7.83 0 0 0 8 480c66.3 0 116-31.8 140.6-51.4A305 305 0 0 0 256 448c141.4 0 256-93.1 256-208S397.4 32 256 32zm32 264a8 8 0 0 1-8 8H136a8 8 0 0 1-8-8v-16a8 8 0 0 1 8-8h144a8 8 0 0 1 8 8zm96-96a8 8 0 0 1-8 8H136a8 8 0 0 1-8-8v-16a8 8 0 0 1 8-8h240a8 8 0 0 1 8 8z"
                fill="currentColor"
              />
              <path
                class="fa-primary"
                d="M376 176H136a8 8 0 0 0-8 8v16a8 8 0 0 0 8 8h240a8 8 0 0 0 8-8v-16a8 8 0 0 0-8-8zm-96 96H136a8 8 0 0 0-8 8v16a8 8 0 0 0 8 8h144a8 8 0 0 0 8-8v-16a8 8 0 0 0-8-8z"
                fill="currentColor"
              />
            </g>
          </svg>
        </fa-icon>
        Kopieer mededeling 
      </span>
      <div
        class="mat-button-ripple mat-ripple"
        matripple=""
      />
      <div
        class="mat-button-focus-overlay"
      />
    </button>
  </div>
</ogm-generate>
`;
